File: app/controllers/sommelier_controller.rb
# frozen_string_literal: true

class SommelierController < ApplicationController
  skip_before_action :authenticate_user!, raise: false
  skip_before_action :set_current_employee, raise: false
  skip_before_action :set_permissions, raise: false
  skip_before_action :redirect_to_onboarding_if_needed, raise: false
  skip_around_action :switch_locale, raise: false

  before_action :set_smartmenu
  before_action :set_menu

  # POST /smartmenus/:smartmenu_id/sommelier/recommend
  def recommend
    preferences = {
      smoky: ActiveModel::Type::Boolean.new.cast(params[:smoky]),
      taste: params[:taste].to_s.presence || 'dry',
      budget: params[:budget].to_i.clamp(1, 3),
    }

    recommender = BeverageIntelligence::Recommender.new
    results = recommender.recommend_for_guest(
      menu: @menu,
      preferences: preferences,
      limit: 3,
    )

    render json: {
      recommendations: results.map { |r| format_recommendation(r) },
      preferences: preferences,
    }
  end

  # POST /smartmenus/:smartmenu_id/sommelier/recommend_wine
  def recommend_wine
    preferences = {
      wine_color: params[:wine_color].to_s.presence || 'no_preference',
      body: params[:body].to_s.presence || 'medium',
      taste: params[:taste].to_s.presence || 'dry',
      budget: params[:budget].to_i.clamp(1, 3),
    }

    recommender = BeverageIntelligence::Recommender.new
    results = recommender.recommend_wines_for_guest(
      menu: @menu,
      preferences: preferences,
      limit: 3,
    )

    render json: {
      recommendations: results.map { |r| format_wine_recommendation(r) },
      preferences: preferences,
    }
  end

  # GET /smartmenus/:smartmenu_id/sommelier/pairings/:menuitem_id
  def pairings
    menuitem = @menu.menuitems.find(params[:menuitem_id])

    pairings = PairingRecommendation
      .where(drink_menuitem_id: menuitem.id)
      .order(score: :desc)
      .limit(4)
      .includes(:food_menuitem)

    similar = []
    link = menuitem.menu_item_product_links.first
    if link
      similar = SimilarProductRecommendation
        .where(product_id: link.product_id)
        .order(score: :desc)
        .limit(3)
        .includes(recommended_product: :menuitems)
    end

    render json: {
      drink: { id: menuitem.id, name: menuitem.name },
      pairings: pairings.map { |p| format_pairing(p) },
      similar: similar.map { |s| format_similar(s, @menu) },
    }
  rescue ActiveRecord::RecordNotFound
    render json: { error: 'Item not found' }, status: :not_found
  end

  # POST /smartmenus/:smartmenu_id/sommelier/recommend_whiskey
  def recommend_whiskey
    preferences = {
      experience_level: params[:experience_level].to_s.presence || 'casual',
      region_pref: params[:region_pref].to_s.presence || 'surprise_me',
      flavor_pref: params[:flavor_pref].to_s.presence,
      budget: params[:budget].to_i.clamp(1, 3),
    }
    exclude_ids = Array(params[:exclude_ids]).map(&:to_i)

    recommender = BeverageIntelligence::WhiskeyRecommender.new
    results = recommender.recommend_for_guest(
      menu: @menu,
      preferences: preferences,
      limit: 3,
      exclude_ids: exclude_ids,
    )

    render json: {
      recommendations: results.map { |r| format_whiskey_recommendation(r) },
      preferences: preferences,
    }
  end

  # GET /smartmenus/:smartmenu_id/sommelier/explore_whiskeys
  def explore_whiskeys
    recommender = BeverageIntelligence::WhiskeyRecommender.new
    result = recommender.explore(
      menu: @menu,
      cluster: params[:cluster].to_s.presence,
      region: params[:region].to_s.presence,
      age_range: params[:age_range].to_s.presence,
      price_range: params[:price_range].to_s.presence,
      new_only: ActiveModel::Type::Boolean.new.cast(params[:new_only]),
      rare_only: ActiveModel::Type::Boolean.new.cast(params[:rare_only]),
    )

    render json: {
      quadrants: result[:quadrants],
      items: result[:items].map { |i| format_explore_item(i) },
    }
  end

  # GET /smartmenus/:smartmenu_id/sommelier/whiskey_flights
  def whiskey_flights
    flights = @menu.whiskey_flights.visible.order(:created_at)

    render json: {
      flights: flights.map { |f| format_flight(f) },
    }
  end

  private

  def set_smartmenu
    @smartmenu = Smartmenu.find_by!(slug: params[:smartmenu_id])
  rescue ActiveRecord::RecordNotFound
    render json: { error: 'Smartmenu not found' }, status: :not_found
  end

  def set_menu
    @menu = @smartmenu.menu
    render json: { error: 'No menu' }, status: :not_found unless @menu
  end

  def format_recommendation(rec)
    item = rec[:menuitem]
    enrichment = rec[:enrichment] || {}
    tasting = enrichment['tasting_notes'] || {}

    result = {
      id: item.id,
      name: item.name,
      description: item.description,
      price: item.price,
      category: item.sommelier_category,
      tags: rec[:tags],
      score: (rec[:score] * 100).round,
    }

    if tasting.any?
      result[:tasting_notes] = tasting.slice('nose', 'palate', 'finish').compact_blank
    end

    result[:story] = enrichment['brand_story'] if enrichment['brand_story'].present?
    result[:region] = enrichment['region'] if enrichment['region'].present?

    if rec[:best_pairing]
      food = rec[:best_pairing].food_menuitem
      result[:best_pairing] = {
        food_name: food.name,
        rationale: rec[:best_pairing].rationale,
        score: rec[:best_pairing].display_score,
      }
    end

    result
  end

  def format_wine_recommendation(rec)
    result = format_recommendation(rec)
    result[:wine_color] = rec[:wine_color] if rec[:wine_color].present?
    result[:grape_variety] = rec[:grape_variety] if rec[:grape_variety].present?
    result[:appellation] = rec[:appellation] if rec[:appellation].present?
    result[:vintage_year] = rec[:vintage_year] if rec[:vintage_year].present?
    result[:classification] = rec[:classification] if rec[:classification].present?
    result
  end

  def format_pairing(pairing)
    {
      food_id: pairing.food_menuitem_id,
      food_name: pairing.food_menuitem.name,
      food_description: pairing.food_menuitem.description,
      food_price: pairing.food_menuitem.price,
      score: pairing.display_score,
      pairing_type: pairing.pairing_type,
      rationale: pairing.rationale,
    }
  end

  def format_similar(rec, menu)
    product = rec.recommended_product
    on_menu_item = product.menuitems
                          .joins(menusection: :menu)
                          .where(menus: { id: menu.id }, status: 'active')
                          .first

    {
      product_name: product.canonical_name,
      score: (rec.score.to_f * 100).round,
      rationale: rec.rationale,
      on_menu: on_menu_item.present?,
      menuitem_id: on_menu_item&.id,
      menuitem_name: on_menu_item&.name,
      menuitem_price: on_menu_item&.price,
    }
  end

  def format_whiskey_recommendation(rec)
    item = rec[:menuitem]
    parsed = rec[:parsed_fields] || {}

    {
      menuitem_id: item.id,
      name: item.name,
      price: item.price,
      whiskey_type: parsed['whiskey_type'],
      region: parsed['whiskey_region'],
      distillery: parsed['distillery'],
      cask_type: parsed['cask_type'],
      age_years: parsed['age_years'],
      abv: parsed['bottling_strength_abv'],
      flavor_cluster: parsed['staff_flavor_cluster'],
      tags: rec[:tags],
      staff_tasting_note: parsed['staff_tasting_note'],
      staff_pick: parsed['staff_pick'] == true,
      why_text: rec[:why_text],
      score: rec[:score],
      new_arrival: rec[:new_arrival],
      rare: rec[:rare],
    }
  end

  def format_explore_item(item_hash)
    item = item_hash[:menuitem]
    parsed = item_hash[:parsed_fields] || {}

    {
      menuitem_id: item.id,
      name: item.name,
      price: item.price,
      whiskey_type: parsed['whiskey_type'],
      region: parsed['whiskey_region'],
      distillery: parsed['distillery'],
      age_years: parsed['age_years'],
      flavor_cluster: item_hash[:cluster],
      tags: item_hash[:tags],
      staff_tasting_note: parsed['staff_tasting_note'],
      staff_pick: parsed['staff_pick'] == true,
      new_arrival: item_hash[:new_arrival],
      rare: item_hash[:rare],
    }
  end

  def format_flight(flight)
    {
      id: flight.id,
      theme_key: flight.theme_key,
      title: flight.title,
      narrative: flight.narrative,
      source: flight.source,
      total_price: flight.total_price,
      display_price: flight.display_price,
      per_dram_price: flight.per_dram_price,
      savings: flight.savings,
      items: flight.items,
    }
  end
end

File: app/javascript/controllers/sommelier_controller.js
import { Controller } from "@hotwired/stimulus"

/**
 * Sommelier Controller — Guest-facing "Help me choose" dual-flow.
 *
 * Supports two flows:
 *   - Spirits: smoky → taste → budget → /recommend
 *   - Wine:    color → style → budget → /recommend_wine
 *
 * Targets:
 *   panel      — the bottom-sheet / overlay container
 *   step       — each step screen (0=mode, 1-3=spirits, 4-6=wine)
 *   results    — results container
 *   loading    — spinner shown during fetch
 */
export default class extends Controller {
  static targets = ["panel", "step", "results", "loading", "footer"]
  static values  = {
    url: String,
    wineUrl: String,
    pairingsUrl: String,
    currency: { type: String, default: "EUR" },
    availability: { type: Object, default: {} },
  }

  connect() {
    this.mode = null // "spirits" or "wine"
    this.preferences = {}
    this.winePreferences = {}
    this.flowSteps = []
  }

  open() {
    this.mode = null
    this.preferences = { smoky: null, taste: null, budget: null }
    this.winePreferences = { wine_color: null, body: null, taste: null, budget: null }

    const avail = this.availabilityValue || {}
    const hasSpirits = (avail.spirits || 0) > 0
    const hasWine = (avail.wine || 0) > 0

    // Hide mode buttons that have zero items
    this._filterModeButtons(hasSpirits, hasWine)

    // Pre-filter wine color buttons
    this._filterWineColorButtons(avail.wine_colors || {})

    // Auto-select mode if only one category exists
    if (hasSpirits && !hasWine) {
      this.mode = "spirits"
      this.flowSteps = this._spiritsStepIndices()
    } else if (hasWine && !hasSpirits) {
      this.mode = "wine"
      this.flowSteps = this._wineStepIndices()
    } else {
      this.flowSteps = [0] // show mode selection
    }

    this.panelTarget.classList.remove("d-none")
    this.panelTarget.classList.add("sommelier-open")
    document.body.classList.add("sommelier-active")
    this._advanceFlow()
  }

  close() {
    this.panelTarget.classList.add("d-none")
    this.panelTarget.classList.remove("sommelier-open")
    document.body.classList.remove("sommelier-active")
  }

  // Step 0: Mode selection
  setMode(event) {
    this.mode = event.currentTarget.dataset.value
    if (this.mode === "wine") {
      this.flowSteps = this._wineStepIndices()
    } else {
      this.flowSteps = this._spiritsStepIndices()
    }
    this._advanceFlow()
  }

  // Hide mode buttons with zero items; if only one visible, auto-select it
  _filterModeButtons(hasSpirits, hasWine) {
    const modeStep = this.stepTargets[0]
    if (!modeStep) return
    const buttons = modeStep.querySelectorAll(".sommelier-option-btn[data-value]")
    buttons.forEach(btn => {
      const val = btn.dataset.value
      if (val === "spirits") btn.classList.toggle("d-none", !hasSpirits)
      if (val === "wine") btn.classList.toggle("d-none", !hasWine)
    })
  }

  // Hide wine color buttons with zero matching wines
  _filterWineColorButtons(wineColors) {
    // Find the wine color step (step 4, first wine flow step)
    const wineColorStep = this.stepTargets.find(el =>
      el.dataset.flow === "wine" && el.querySelector("[data-action*='setWineColor']")
    )
    if (!wineColorStep) return
    const buttons = wineColorStep.querySelectorAll(".sommelier-option-btn")
    buttons.forEach(btn => {
      const color = btn.dataset.value
      if (!color || color === "no_preference") return // always show "Surprise me"
      const count = wineColors[color] || 0
      btn.classList.toggle("d-none", count === 0)
    })
  }

  // === SPIRITS FLOW ===

  setSmoky(event) {
    this.preferences.smoky = event.currentTarget.dataset.value === "true"
    this._advanceFlow()
  }

  setTaste(event) {
    this.preferences.taste = event.currentTarget.dataset.value
    this._advanceFlow()
  }

  setBudget(event) {
    this.preferences.budget = parseInt(event.currentTarget.dataset.value, 10)
    this._advanceFlow()
  }

  // === WINE FLOW ===

  setWineColor(event) {
    this.winePreferences.wine_color = event.currentTarget.dataset.value
    this._advanceFlow()
  }

  setWineStyle(event) {
    this.winePreferences.body = event.currentTarget.dataset.body
    this.winePreferences.taste = event.currentTarget.dataset.taste
    this._advanceFlow()
  }

  setWineBudget(event) {
    this.winePreferences.budget = parseInt(event.currentTarget.dataset.value, 10)
    this._advanceFlow()
  }

  // === FLOW NAVIGATION ===

  _spiritsStepIndices() {
    return this.stepTargets
      .map((el, i) => ({ el, i }))
      .filter(({ el }) => el.dataset.flow === "spirits")
      .map(({ i }) => i)
  }

  _wineStepIndices() {
    return this.stepTargets
      .map((el, i) => ({ el, i }))
      .filter(({ el }) => el.dataset.flow === "wine")
      .map(({ i }) => i)
  }

  _advanceFlow() {
    if (this.flowSteps.length === 0) {
      this._fetchResults()
      return
    }
    const nextIdx = this.flowSteps.shift()

    // Auto-skip steps where only one option is visible
    const step = this.stepTargets[nextIdx]
    if (step) {
      const visibleButtons = Array.from(step.querySelectorAll(".sommelier-option-btn"))
        .filter(btn => !btn.classList.contains("d-none"))
      if (visibleButtons.length === 1) {
        visibleButtons[0].click()
        return
      }
    }

    this.showStep(nextIdx)
  }

  showStep(index) {
    this.stepTargets.forEach((el, i) => {
      el.classList.toggle("d-none", i !== index)
    })
    if (this.hasResultsTarget) this.resultsTarget.classList.add("d-none")
    if (this.hasFooterTarget) this.footerTarget.classList.add("d-none")
  }

  async _fetchResults() {
    this.stepTargets.forEach(el => el.classList.add("d-none"))
    if (this.hasLoadingTarget) this.loadingTarget.classList.remove("d-none")
    if (this.hasResultsTarget) this.resultsTarget.classList.add("d-none")
    if (this.hasFooterTarget) this.footerTarget.classList.add("d-none")

    const isWine = this.mode === "wine"
    const url = isWine ? this.wineUrlValue : this.urlValue
    const body = isWine ? this.winePreferences : this.preferences

    try {
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "X-CSRF-Token": this.csrfToken,
        },
        body: JSON.stringify(body),
      })

      if (!response.ok) throw new Error(`HTTP ${response.status}`)

      const data = await response.json()
      this.renderResults(data.recommendations, isWine)
    } catch (error) {
      console.error("[Sommelier] Recommendation failed:", error)
      this.renderError()
    } finally {
      if (this.hasLoadingTarget) this.loadingTarget.classList.add("d-none")
    }
  }

  renderResults(recommendations, isWine = false) {
    if (!this.hasResultsTarget) return
    this.resultsTarget.classList.remove("d-none")

    if (!recommendations || recommendations.length === 0) {
      this.resultsTarget.innerHTML = `
        <div class="sommelier-empty">
          <p class="sommelier-empty-text">No recommendations found for your preferences. Try different options!</p>
        </div>
      `
      if (this.hasFooterTarget) {
        this.footerTarget.classList.remove("d-none")
        this.footerTarget.innerHTML = `
          <button class="sommelier-btn sommelier-btn-outline" data-action="click->sommelier#restart">Try again</button>
        `
      }
      return
    }

    const title = isWine ? "Your Wine Picks" : "Your Picks"
    const cards = recommendations.map((rec, i) =>
      isWine ? this.buildWineCard(rec, i) : this.buildCard(rec, i)
    ).join("")

    this.resultsTarget.innerHTML = `
      <div class="sommelier-results-header">
        <h3 class="sommelier-results-title">${title}</h3>
        <p class="sommelier-results-subtitle">Based on your preferences</p>
      </div>
      <div class="sommelier-cards">${cards}</div>
    `

    if (this.hasFooterTarget) {
      this.footerTarget.classList.remove("d-none")
      this.footerTarget.innerHTML = `
        <button class="sommelier-btn sommelier-btn-outline" data-action="click->sommelier#restart">
          Try different preferences
        </button>
      `
    }
  }

  buildCard(rec, index) {
    const medal = index === 0 ? "🥇" : index === 1 ? "🥈" : "🥉"
    const tags = (rec.tags || []).map(t =>
      `<span class="sommelier-tag">${t.replace(/_/g, " ")}</span>`
    ).join("")

    const tasting = rec.tasting_notes
      ? `<div class="sommelier-tasting">
           ${rec.tasting_notes.nose ? `<div><strong>Nose:</strong> ${rec.tasting_notes.nose}</div>` : ""}
           ${rec.tasting_notes.palate ? `<div><strong>Palate:</strong> ${rec.tasting_notes.palate}</div>` : ""}
           ${rec.tasting_notes.finish ? `<div><strong>Finish:</strong> ${rec.tasting_notes.finish}</div>` : ""}
         </div>`
      : ""

    const pairing = rec.best_pairing
      ? `<div class="sommelier-pairing">
           <span class="sommelier-pairing-label">Pairs with:</span>
           <span class="sommelier-pairing-food">${rec.best_pairing.food_name}</span>
         </div>`
      : ""

    const story = rec.story
      ? `<p class="sommelier-story">${rec.story}</p>`
      : ""

    const region = rec.region
      ? `<span class="sommelier-region">${rec.region}</span>`
      : ""

    return `
      <div class="sommelier-card" data-menuitem-id="${rec.id}">
        <div class="sommelier-card-header">
          <span class="sommelier-medal">${medal}</span>
          <div class="sommelier-card-title-wrap">
            <h4 class="sommelier-card-title">${rec.name}</h4>
            ${region}
          </div>
          <span class="sommelier-card-price">${rec.price ? this.formatPrice(rec.price) : ""}</span>
        </div>
        ${rec.description ? `<p class="sommelier-card-desc">${rec.description}</p>` : ""}
        <div class="sommelier-tags">${tags}</div>
        ${tasting}
        ${story}
        ${pairing}
        <div class="sommelier-card-footer">
          <span class="sommelier-match-score">${rec.score}% match</span>
          ${rec.has_pairings ? `<button class="sommelier-btn sommelier-btn-sm"
                  data-action="click->sommelier#showPairings"
                  data-menuitem-id="${rec.id}">
            View pairings
          </button>` : ""}
        </div>
      </div>
    `
  }

  buildWineCard(rec, index) {
    const medal = index === 0 ? "🥇" : index === 1 ? "🥈" : "🥉"
    const tags = (rec.tags || []).map(t =>
      `<span class="sommelier-tag">${t.replace(/_/g, " ")}</span>`
    ).join("")

    // Wine-specific metadata line
    const wineMeta = []
    if (rec.wine_color) wineMeta.push(rec.wine_color.charAt(0).toUpperCase() + rec.wine_color.slice(1))
    if (rec.grape_variety) wineMeta.push(rec.grape_variety)
    if (rec.appellation) wineMeta.push(rec.appellation)
    if (rec.vintage_year) wineMeta.push(rec.vintage_year)
    if (rec.classification) wineMeta.push(rec.classification)
    const wineMetaHtml = wineMeta.length > 0
      ? `<div class="sommelier-wine-meta">${wineMeta.join(" · ")}</div>`
      : ""

    const pairing = rec.best_pairing
      ? `<div class="sommelier-pairing">
           <span class="sommelier-pairing-label">Pairs with:</span>
           <span class="sommelier-pairing-food">${rec.best_pairing.food_name}</span>
         </div>`
      : ""

    const story = rec.story
      ? `<p class="sommelier-story">${rec.story}</p>`
      : ""

    return `
      <div class="sommelier-card sommelier-card-wine" data-menuitem-id="${rec.id}">
        <div class="sommelier-card-header">
          <span class="sommelier-medal">${medal}</span>
          <div class="sommelier-card-title-wrap">
            <h4 class="sommelier-card-title">${rec.name}</h4>
            ${wineMetaHtml}
          </div>
          <span class="sommelier-card-price">${rec.price ? this.formatPrice(rec.price) : ""}</span>
        </div>
        ${rec.description ? `<p class="sommelier-card-desc">${rec.description}</p>` : ""}
        <div class="sommelier-tags">${tags}</div>
        ${story}
        ${pairing}
        <div class="sommelier-card-footer">
          <span class="sommelier-match-score">${rec.score}% match</span>
          ${rec.has_pairings ? `<button class="sommelier-btn sommelier-btn-sm"
                  data-action="click->sommelier#showPairings"
                  data-menuitem-id="${rec.id}">
            View pairings
          </button>` : ""}
        </div>
      </div>
    `
  }

  async showPairings(event) {
    const itemId = event.currentTarget.dataset.menuitemId
    const url = this.pairingsUrlValue.replace(":menuitem_id", itemId)

    try {
      const response = await fetch(url, {
        headers: { "Accept": "application/json" },
      })
      if (!response.ok) throw new Error(`HTTP ${response.status}`)

      const data = await response.json()
      this.renderPairingsModal(data)
    } catch (error) {
      console.error("[Sommelier] Pairings fetch failed:", error)
    }
  }

  renderPairingsModal(data) {
    const existing = document.getElementById("sommelier-pairings-modal")
    if (existing) existing.remove()

    const pairingsHtml = (data.pairings || []).map(p => `
      <div class="sommelier-pairing-card">
        <div class="sommelier-pairing-card-header">
          <h5>${p.food_name}</h5>
          <span class="sommelier-match-score">${p.score}%</span>
        </div>
        ${p.food_description ? `<p class="sommelier-pairing-desc">${p.food_description}</p>` : ""}
        ${p.food_price ? `<span class="sommelier-pairing-price">${this.formatPrice(p.food_price)}</span>` : ""}
        <p class="sommelier-pairing-rationale">${p.rationale || ""}</p>
        ${p.pairing_type === "surprise" ? '<span class="sommelier-badge-surprise">Surprise pick</span>' : ""}
      </div>
    `).join("")

    const similarHtml = (data.similar || []).filter(s => s.on_menu).map(s => `
      <div class="sommelier-similar-card">
        <h5>${s.menuitem_name || s.product_name}</h5>
        <span class="sommelier-match-score">${s.score}% similar</span>
        ${s.menuitem_price ? `<span class="sommelier-pairing-price">${this.formatPrice(s.menuitem_price)}</span>` : ""}
        <p class="sommelier-similar-rationale">${s.rationale || ""}</p>
      </div>
    `).join("")

    const modal = document.createElement("div")
    modal.id = "sommelier-pairings-modal"
    modal.className = "sommelier-modal-overlay"
    modal.innerHTML = `
      <div class="sommelier-modal">
        <div class="sommelier-modal-header">
          <h3>Pairings for ${data.drink.name}</h3>
          <button class="sommelier-modal-close" id="sommelier-pairings-close-btn">✕</button>
        </div>
        <div class="sommelier-modal-body">
          ${pairingsHtml.length > 0 ? `<h4>Food pairings</h4>${pairingsHtml}` : "<p>No pairings available yet.</p>"}
          ${similarHtml.length > 0 ? `<h4 class="mt-3">You might also like</h4>${similarHtml}` : ""}
        </div>
      </div>
    `
    document.body.appendChild(modal)

    // Wire close handlers directly (modal is outside Stimulus scope)
    modal.querySelector("#sommelier-pairings-close-btn").addEventListener("click", () => this.closePairingsModal())
    modal.addEventListener("click", (e) => { if (e.target === modal) this.closePairingsModal() })
  }

  closePairingsModal() {
    const modal = document.getElementById("sommelier-pairings-modal")
    if (modal) modal.remove()
  }

  renderError() {
    if (!this.hasResultsTarget) return
    this.resultsTarget.classList.remove("d-none")
    this.resultsTarget.innerHTML = `
      <div class="sommelier-empty">
        <p class="sommelier-empty-text">Something went wrong. Please try again.</p>
      </div>
    `
    if (this.hasFooterTarget) {
      this.footerTarget.classList.remove("d-none")
      this.footerTarget.innerHTML = `
        <button class="sommelier-btn sommelier-btn-outline" data-action="click->sommelier#restart">Retry</button>
      `
    }
  }

  restart() {
    this.open()
  }

  formatPrice(price) {
    if (!price) return ""
    try {
      return parseFloat(price).toLocaleString(undefined, {
        style: "currency",
        currency: this.currencyValue,
        minimumFractionDigits: 2,
      })
    } catch {
      return `${parseFloat(price).toFixed(2)} ${this.currencyValue}`
    }
  }

  get csrfToken() {
    const meta = document.querySelector('meta[name="csrf-token"]')
    return meta ? meta.content : ""
  }
}
